### PR TITLE
DEPR: DatetimeIndex indexing with mismatched tzawareness

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: doc/source/whatsnew/v2.0.0.rst [hunk 1]
```diff
@@ -302,6 +302,7 @@ Removal of prior version deprecations/changes
 - Changed behavior of empty data passed into :class:`Series`; the default dtype will be ``object`` instead of ``float64`` (:issue:`29405`)
 - Changed the behavior of :meth:`DatetimeIndex.union`, :meth:`DatetimeIndex.intersection`, and :meth:`DatetimeIndex.symmetric_difference` with mismatched timezones to convert to UTC instead of casting to object dtype (:issue:`39328`)
 - Changed the behavior of :func:`to_datetime` with argument "now" with ``utc=False`` to match ``Timestamp("now")`` (:issue:`18705`)
+- Changed the behavior of indexing on a timezone-aware :class:`DatetimeIndex` with a timezone-naive ``datetime`` object or vice-versa; these now behave like any other non-comparable type (:issue:`36148`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.0.0.rst:305]: Might be good to clarify `behave like any other non-comparable type` means raising a `KeyError`?

File: pandas/core/indexes/datetimes.py [hunk 2]
```diff
@@ -552,31 +552,24 @@ def _parsed_string_to_bounds(self, reso: Resolution, parsed: datetime):
...
+    def _disallow_mismatched_indexing(self, key, one_way: bool = False) -> None:
+        """
+        Check for mismatched-tzawareness indexing and re-raise as KeyError.
+        """
+        try:
+            self._deprecate_mismatched_indexing(key, one_way=one_way)
+        except TypeError as err:
+            raise KeyError(key) from err
+
     def _deprecate_mismatched_indexing(self, key, one_way: bool = False) -> None:
```
Comment (mroeschke) [inline pandas/core/indexes/datetimes.py:564]: Is this function still used elsewhere and cannot be rolled into `_disallow_mismatched_indexing`?

Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py:564]: its still used in `_maybe_cast_slice_bound`, where I think TypeError makes more sense than KeyError

File: doc/source/whatsnew/v2.0.0.rst (hunk 1 above)
Comment (jbrockmendel) [reply doc/source/whatsnew/v2.0.0.rst]: good idea, updated

**Review by mroeschke — COMMENTED** [review]
Looks good. Just a merge conflict

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

**Pull request merged**